src/main/kotlin/com/exactpro/th2/replay/script/generator/MessageTransformer.kt

```diff
@@ -19,105 +19,120 @@ package com.exactpro.th2.replay.script.generator
         }
     }
 
-    private fun getTransform(message: Message) = transforms[message.metadata.protocol]?.get(message.messageType)
+    private val Message.commands
+        get() = transforms[metadata.protocol]?.get(messageType)
 
     companion object {
         private val LOGGER = KotlinLogging.logger {}
+        private val PATH_CONFIG = Configuration.builder().options(AS_PATH_LIST).build()
 
-        private fun <T> T.transform(path: List<String>, transforms: FieldTransformations) {
-            when (this) {
-                is MutableList<*> -> (this as MutableList<Any?>).transform(path, transforms)
-                is MutableMap<*, *> -> (this as MutableMap<String, Any?>).transform(path, transforms)
-            }
+        private fun Map<String, Any?>.getMatchingPaths(path: JsonPath): List<JsonPath> = try {
+            path.read<List<String>>(this, PATH_CONFIG).map(JsonPath::compile)
+        } catch (e: PathNotFoundException) {
+            listOf()
         }
 
-        private fun MutableMap<String, Any?>.transform(path: List<String>, fields: FieldTransformations) {
-            val iterator = iterator()
-
-            while (iterator.hasNext()) {
-                val entry = iterator.next()
-                val subPath = path + entry.key
-
-                fields.entries.firstOrNull { (path, _) ->
-                    path.matches(subPath)
-                }?.let { (_, transform) ->
-                    when (transform.action) {
-                        REMOVE -> iterator.remove().also {
-                            LOGGER.debug { "Removed value '$it', path '$subPath'" }
-                        }
-                        REPLACE -> entry.setValue(transform.value).also {
-                            LOGGER.debug { "Replaced from '$it' to '${transform.value}', path '$subPath'" }
-                        }
-                        CHECK_AND_REPLACE -> if (entry.value == transform.expected) {
-                            entry.setValue(transform.value)
-                            LOGGER.debug { "Replaced from '${transform.expected}' to '${transform.value}', path '$subPath'" }
-                        } else {
-                            LOGGER.debug { "Transformation '$transform' skipped because expected value didn't match actual '${entry.value}', path $subPath" }
-                        }
-                    }
-                } ?: run {
-                    entry.value.transform(subPath, fields)
+        private operator fun DocumentContext.get(path: JsonPath): Any? = try {
+            read<Any?>(path)
+        } catch (e: PathNotFoundException) {
+            null
+        }
+
+        private fun JsonPath.resolve(path: JsonPath?): JsonPath = when {
+            path == null -> this
+            path.path[0] == '@' -> JsonPath.compile(this.path + path.path.drop(1))
+            else -> path
+        }
+
+        fun MutableMap<String, out Any?>.apply(commands: List<Command>) = apply {
```

Comment:
Is it OK to have overloads for apply method? Maybe we should call it `applyCommands`? When I read the code it was quite hard to understand that it is a custom `apply` method

Reply:
I don't see any problems with it, but I can rename it if you want

Reply:
Just for better readability without IDE, I would prefer if you do. Thanks